pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,205 @@
+	PostBind6:    ebpf.AttachCGroupInet6PostBind,
+	PreBind6:     ebpf.AttachCGroupInet6Bind,
+}
+
+// attachCgroup attaches a program from spec with the given name to cgroupRoot.
+// If the kernel supports it, the resulting bpf_link is pinned to pinPath.
+//
+// Upgrades from prior Cilium versions will continue to be handled by a PROG_ATTACH
+// to replace an old program attached to a cgroup.
+func attachCgroup(spec *ebpf.Collection, name, cgroupRoot, pinPath string) error {
+	prog := spec.Programs[name]
+	if prog == nil {
+		return fmt.Errorf("program %s not found in ELF", name)
+	}
+
+	// A previously created link is found and updated.
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Update(prog); err != nil {
+			return fmt.Errorf("update link %s: %w", pinPath, err)
+		}
+
+		log.Debugf("updated link %s for program %s", pinPath, name)
+		return nil
+	}
+	log.Debugf("no pinned link at '%s'", pinPath)
+
+	cg, err := os.Open(cgroupRoot)
+	if err != nil {
+		return fmt.Errorf("open cgroup %s: %w", cgroupRoot, err)
+	}
+	defer cg.Close()
+
+	// no previous link is found, try to create one. This will only work on fresh installs.
+	l, err = link.AttachRawLink(link.RawLinkOptions{
+		Target:  int(cg.Fd()),
+		Program: prog,
+		Attach:  attachTypes[name],
+	})
+
+	switch {
+	// Hit a recoverable error that could be resolved by attaching the prog
+	// with PROG_ATTACH.
+	//
+	// Kernels before 5.7 don't support bpf_link.
+	// In that case link.AttachRawLink returns ErrNotSupported.
+	//
+	// If the kernel supports bpf_link, and an older version of Cilium attached a
+	// cgroup program without flags (old init.sh behaviour), link.AttachRawLink
+	// will return EPERM because bpf_link implicitly uses the multi flag.
+	case errors.Is(err, unix.EPERM) || errors.Is(err, link.ErrNotSupported):
+		log.Debugf("performing fallback attach for program %s", name)
+
+		err = link.RawAttachProgram(link.RawAttachProgramOptions{
+			Target:  int(cg.Fd()),
+			Program: prog,
+			Attach:  attachTypes[name],
+		})
+		if err != nil {
+			return fmt.Errorf("fallback attach program %s: %w", name, err)
+		}
+
+		// Nothing left to do, the cgroup now holds a reference to the prog
+		// so we don't need to hold a reference in the agent/bpffs to ensure
+		// the program stays active.
+		log.Debugf("program %s was attached using PROG_ATTACH", name)
+		return nil
+
+	case err != nil:
+		return fmt.Errorf("attach program %s: %w", name, err)
+	}
+
+	// The program was successfully attached using bpf_link, and there was no preexisting pinned link.
+	// The link needs to be pinned and closed.
+	err = l.Pin(filepath.Join(pinPath, name))
+	if err != nil {
+		return fmt.Errorf("pin program %s link at %s: %w", name, pinPath, err)
+	}
+
+	// Successfully created and pinned bpf_link. Closing the link will not remove it, as it is pinned.
+	log.Debugf("program %s attached using bpf_link", name)
+	return l.Close()
```

Comment:
nit: If the close fails, how does that happen and does that really make sense to return the error? My point being, it looks like the program was loaded, attached and pinned - so the functionality of this call is achieved successfully at this point. Even if there's a close fail, is there anything the caller should do about it? May consider just dropping a info/warning log and returning nil at this point.

Reply:
I agree that returning it like that is wrong, what about just defering the Close? Otherwise I'll go with calling Close and logging potential errors.

Reply:
Either way. The common pattern is:

```
foo, err := TrySomething()
if err != nil {
    // handle err
}
defer func() {
    if err := foo.Close(); err != nil {
        // log at warning level (?)
    } 
}
```

But if I understand correctly, there's a deliberate choice here *not* to close the link if pinning fails. So I guess it could just be open-coded at the end of the function here?